backend/controllers/userController.js

```diff
@@ -1,43 +1,142 @@
-
-    if (userExists) {
-      res.status(400)
-      throw new Error(`User already exists`)
+    const { userName, name, email, password, passwordVerify } = req.body;
+    //validation
+    //validate that it's all here!
+    if(!userName || !name || !email || !password || !passwordVerify) {
+      // respond with 400 which is a bad request and json object with a key: errorMessage and a response: the message
+      return res.status(400).json({ errorMessage: `Please fill in all fields` })
+    }
+    //validate password length
+    if(password.length < 6) {
+      return res.status(400).json({ errorMessage: `Password must be at least 6 characters` })
     }
-    // user.create is just syntactic sugar for user.save
-    const user = await User.create({
+    //validate that password and verify password match
+    if(password !== passwordVerify) {
+      return res.status(400).json({ errorMessage: `Passwords don't match` })
+    }
+    // validate that email is unique, user.findOne returns a promise, so we need to await it
+    const userExists = await User.findOne({ email })
+    if(userExists) {
+      return res.status(400).json({ errorMessage: `This email is already in use` })
+    } 
+    // we encript the password in the User model not here
+    // create a new User, User.create is a method / shorthand that creates, then saves our user to the db
+    const newUser = await User.create({
+      userName,
       name, 
-      userName, 
+      email, 
       password
     })
 
-    if(user) {
+    // log the user in! we say if data.token exists && matches our cookie, then user is logged in :D
+    if(newUser) {
+      const token = generateToken(newUser._id)
+      // send token in a cookie
+      res.cookie('token', token, {
+        httpOnly: true,
+        secure: true,
+        sameSite: "none",
+        MaxAge: 36000
+      })
       res.status(201).json({
+        _id: newUser._id,
+        userName: newUser.userName,
+        name: newUser.name,
+        email: newUser.email,
+        token: token
+      })
+    } else {
+      res.status(400).json({ errorMessage: `Invalid user data` })
+    }
+ 
+  } catch (err) {
+    console.error(err.message);
+    //send the status and the message 500 = internal server error, send a message so others can't see anything specific
+    res.status(500).json({ errorMessage: `Server error` })
+  }
+
+}
+
+
+// @desc    log user in & get token
+// @route   POST /api/users/login
+// @access  Public
+const logginUser = async (req, res) => {
+  try {
+    const { email, password } = req.body
+     //validation
+    //validate that it's all here!
+    if( !email || !password ) {
+      // respond with 400 which is a bad request and json object with a key: errorMessage and a response: the message
+      return res.status(400).json({ errorMessage: `Please fill in all fields` })
+    }
+    
+    const user = await User.findOne({ email })
+
+    if (user && (await user.matchPassword(password))) {
```

Comment:
I would prefer you don't have an await in the `if` statement.
Better to save the result in a variable and that variable you can use in the conditional statement